include/lbann/layers/layer.hpp

```diff
@@ -101,6 +108,11 @@ class Layer {
    *  human-readable, name.
    */
   inline void set_name(const std::string name) { m_name = name; }
+  /** Get a string representing the layer datatype
+   */
+  virtual std::string get_datatype_name() const {
```

Comment:
Logically speaking, it doesn't make sense for the base layer class to have a single datatype. For example, what about a convolution layer that takes in `half`s and outputs `float`s? It does make sense for `data_type_layer`s to have a single type, since those explicitly have one type for their inputs and outputs.

It's not a major thing, but it does muddle the logical notion of an abstract layer.

Reply:
So are you thinking that this should be something like `get_weights_datatype_name` or `get_activations_datatype_name`

Reply:
I think that we can clarify this when we get to multiple datatypes per layer.

Reply:
Just my 2 cents: I think this should wait for a future PR. As @bvanessen notes, there's only one datatype currently associated with the layer. There might be multiple input layers to a given layer, and I don't know that their activation types should be directly accessible from "this" layer. When we split the activations type and the weights type, we can change the API.